Index: Classes/Mail/TransportFactory.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Classes/Mail/TransportFactory.php b/Classes/Mail/TransportFactory.php
--- a/Classes/Mail/TransportFactory.php	
+++ b/Classes/Mail/TransportFactory.php	(date 1696934375000)
@@ -74,8 +74,8 @@
         }
 
         $transportType = isset($mailSettings['transport_spool_type'])
-            && !empty($mailSettings['transport_spool_type'])
-            ? 'spool' : (string)$mailSettings['transport'];
+        && !empty($mailSettings['transport_spool_type'])
+            ? 'spool' : $mailSettings['transport'];
 
         switch ($transportType) {
             case 'spool':
@@ -98,7 +98,7 @@
                 if ($host === '') {
                     throw new Exception('$GLOBALS[\'TYPO3_CONF_VARS\'][\'MAIL\'][\'transport_smtp_server\'] needs to be set when transport is set to "smtp".', 1291068606);
                 }
-                if ($port === null) {
+                if ($port === null || $port === '') {
                     $port = 25;
                 } else {
                     $port = (int)$port;
@@ -112,6 +112,11 @@
                     $this->dispatcher,
                     $this->logManager->getLogger(EsmtpTransport::class)
                 );
+                $streamOptions = (array)($mailSettings['transport_smtp_stream_options'] ?? []);
+                if (!empty($streamOptions)) {
+                    $stream = $transport->getStream();
+                    $stream->setStreamOptions(array_merge_recursive($stream->getStreamOptions(), $streamOptions));
+                }
                 // Need authentication?
                 $username = (string)($mailSettings['transport_smtp_username'] ?? '');
                 if ($username !== '') {
@@ -125,6 +130,19 @@
                 if ($mailDomain !== '') {
                     $transport->setLocalDomain($mailDomain);
                 }
+                $restartThreshold = (int)($mailSettings['transport_smtp_restart_threshold'] ?? 0);
+                $restartThresholdSleep = (int)($mailSettings['transport_smtp_restart_threshold_sleep'] ?? 0);
+                if ($restartThreshold > 0) {
+                    if ($restartThresholdSleep < 0) {
+                        // invalid, use default for threshold sleep
+                        $restartThresholdSleep = 0;
+                    }
+                    $transport->setRestartThreshold($restartThreshold, $restartThresholdSleep);
+                }
+                $pingThreshold = (int)($mailSettings['transport_smtp_ping_threshold'] ?? 0);
+                if ($pingThreshold > 0) {
+                    $transport->setPingThreshold($pingThreshold);
+                }
                 break;
             case 'sendmail':
                 $sendmailCommand = $mailSettings['transport_sendmail_command'] ?? @ini_get('sendmail_path');
@@ -140,8 +158,8 @@
                 );
                 break;
             case 'mbox':
-                $mboxFile = (string)($mailSettings['transport_mbox_file'] ?? '');
-                if ($mboxFile === '') {
+                $mboxFile = $mailSettings['transport_mbox_file'];
+                if ($mboxFile == '') {
                     throw new Exception('$GLOBALS[\'TYPO3_CONF_VARS\'][\'MAIL\'][\'transport_mbox_file\'] needs to be set when transport is set to "mbox".', 1294586645);
                 }
                 // Create our transport
@@ -152,7 +170,7 @@
                     $this->logManager->getLogger(MboxTransport::class)
                 );
                 break;
-                // Used for testing purposes
+            // Used for testing purposes
             case 'null':
             case NullTransport::class:
                 $transport = new NullTransport(
@@ -160,7 +178,7 @@
                     $this->logManager->getLogger(NullTransport::class)
                 );
                 break;
-                // Used by Symfony's Transport Factory
+            // Used by Symfony's Transport Factory
             case !empty($mailSettings['dsn']):
             case 'dsn':
                 if (empty($mailSettings['dsn'])) {
@@ -181,6 +199,7 @@
                             but must implement that interface to be used as a mail transport.', 1323006478);
                 }
         }
+
         return $transport;
     }
 
@@ -193,10 +212,9 @@
      */
     protected function createSpool(array $mailSettings): DelayedTransportInterface
     {
-        $transportSpoolType = (string)($mailSettings['transport_spool_type'] ?? '');
-        switch ($transportSpoolType) {
+        switch ($mailSettings['transport_spool_type']) {
             case self::SPOOL_FILE:
-                $path = GeneralUtility::getFileAbsFileName($mailSettings['transport_spool_filepath'] ?? '');
+                $path = GeneralUtility::getFileAbsFileName($mailSettings['transport_spool_filepath']);
                 if (empty($path)) {
                     throw new \RuntimeException('The Spool Type filepath must be configured for TYPO3 in order to be used. Be sure that it\'s not accessible via the web.', 1518558797);
                 }
@@ -215,7 +233,7 @@
                 );
                 break;
             default:
-                $spool = GeneralUtility::makeInstance($transportSpoolType, $mailSettings);
+                $spool = GeneralUtility::makeInstance($mailSettings['transport_spool_type'], $mailSettings);
                 if (!($spool instanceof DelayedTransportInterface)) {
                     throw new \RuntimeException(
                         $mailSettings['transport_spool_type'] . ' is not an implementation of DelayedTransportInterface, but must implement that interface to be used as a mail spool.',
@@ -226,4 +244,4 @@
         }
         return $spool;
     }
-}
+}
\ No newline at end of file
